extmod/machine_i2c_target.c

```diff
@@ -0,0 +1,356 @@
+}
+
+static void machine_i2c_target_data_init(machine_i2c_target_data_t *data, mp_obj_t mem_obj, mp_int_t mem_addrsize) {
+    data->state = STATE_IDLE;
+    data->mem_addr = 0;
+    data->mem_len = 0;
+    data->mem_buf = NULL;
+    if (mem_obj != mp_const_none) {
+        mp_buffer_info_t bufinfo;
+        mp_get_buffer_raise(mem_obj, &bufinfo, MP_BUFFER_RW);
+        if (mem_addrsize < 8 || mem_addrsize > 32 || mem_addrsize % 8 != 0) {
+            mp_raise_ValueError(MP_ERROR_TEXT("mem_addrsize must be 8, 16, 24 or 32"));
+        }
+        data->mem_len = bufinfo.len;
+        data->mem_buf = bufinfo.buf;
+    }
+}
+
+static void machine_i2c_target_data_reset_helper(machine_i2c_target_data_t *data) {
+    size_t len = 0; // TODO
+    if (data->state == STATE_READING) {
+        event(data, I2C_TARGET_IRQ_READ_END, data->mem_addr, len);
+    } else if (data->state == STATE_WRITING) {
+        event(data, I2C_TARGET_IRQ_WRITE_END, data->mem_addr, len);
+    }
+    data->state = STATE_IDLE;
+}
+
+static void machine_i2c_target_data_addr_match(machine_i2c_target_data_t *data, bool read) {
+    event(data, I2C_TARGET_IRQ_ADDR_MATCH, read, 0);
+    machine_i2c_target_data_reset_helper(data);
+}
+
+static void machine_i2c_target_data_read_request(machine_i2c_target_obj_t *self, machine_i2c_target_data_t *data) {
+    // Let the user handle the read request.
+    bool event_handled = event(data, I2C_TARGET_IRQ_READ_REQ, 0, 0);
+    if (data->mem_buf == NULL) {
+        if (!event_handled) {
+            // No data source, just write out a zero.
+            uint8_t val = 0;
+            mp_machine_i2c_target_write_bytes(self, 1, &val);
+        }
+    } else {
+        // Have a buffer.
+        if (data->state == STATE_MEM_ADDR_SELECTED) {
+            data->state = STATE_READING;
+            event(data, I2C_TARGET_IRQ_READ_START, data->mem_addr, 0);
+        }
+        uint8_t val = data->mem_buf[data->mem_addr++];
+        if (data->mem_addr >= data->mem_len) {
+            data->mem_addr = 0;
+        }
+        mp_machine_i2c_target_write_bytes(self, 1, &val);
+    }
+}
+
+static void machine_i2c_target_data_write_request(machine_i2c_target_obj_t *self, machine_i2c_target_data_t *data) {
+    // Let the user handle the write request.
+    bool event_handled = event(data, I2C_TARGET_IRQ_WRITE_REQ, 0, 0);
+    if (data->mem_buf == NULL) {
+        if (!event_handled) {
+            // No data sink, just read and discard the incoming byte.
+            uint8_t buf;
+            mp_machine_i2c_target_read_bytes(self, 1, &buf);
+        }
+    } else {
+        // Have a buffer.
+        uint8_t val;
+        mp_machine_i2c_target_read_bytes(self, 1, &val);
+        if (data->state == STATE_IDLE) {
+            // TODO allow N bytes for address, with N=0 allowed
+            data->state = STATE_MEM_ADDR_SELECTED;
+            data->mem_addr = val % data->mem_len;
+            event(data, I2C_TARGET_IRQ_MEM_ADDR_MATCH, data->mem_addr, 0);
```

Comment:
Seems to cause a duplicate call, because it follows the address byte handling, which already calls the event handler for I2C_TARGET_IRQ_MEM_ADDR_MATCH.